### PR TITLE
Add support for `delta` and `creduce`

### PR DESCRIPTION
**Pull request by tamird**
Closes #34220.

cc @erickt

### CODE REVIEW

File: Library/Formula/creduce.rb [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+    system "./configure", "--prefix=#{prefix}",
+                          "--disable-dependency-tracking",
+                          "--with-llvm=#{llvm_prefix}",
+                          "--bindir=#{libexec}"
+    system "make", "install"
+
+    (bin + "creduce").write_env_script("#{libexec}/creduce", :PERL5LIB => ENV["PERL5LIB"])
+  end
+end
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Can a test be added to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

File: Library/Formula/delta.rb [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+# Two requests are needed to download from Tigris.org. In the first, the server
+# provides a cookie and then redirects the client to a servlet. The servlet
+# requires this cookie, or it will send the client in a redirect loop.
+#
+# The Homebrew CurlDownloadStrategy does not store cookies, so this subclass
+# adds the -c flag and a temporary cookie storage file.
+#
+# This relies on internal details of the CurlDownloadStrategy, so expect it to
+# break.
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: We can't include this in this state then, sorry.

File: Library/Formula/delta.rb [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+  sha1 "e5ab4933bdbcddac3ba96d6ec497213b42f9f26e"
+
+  head "http://delta.tigris.org/svn/delta/trunk",
+    :using => GuestSubversionDownloadStrategy
+
+  def install
+    system "make"
+    bin.install("delta")
+    bin.install("multidelta")
+    bin.install("topformflat")
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: We should use a `make install` if there is one.

**Comment by mikemcquaid** [comment]
A bunch of failing tests that need fixed: http://bot.brew.sh/job/Homebrew%20Pull%20Requests/24277/

File: Library/Formula/delta.rb (hunk 2 above)
Comment (tamird) [inline Library/Formula/delta.rb]: any suggestions on how to deal with this?

Comment (mikemcquaid) [inline Library/Formula/delta.rb]: Find a mirror which doesn't require these weird download approach, I guess?

File: Library/Formula/delta.rb (hunk 3 above)
Comment (tamird) [inline Library/Formula/delta.rb]: there is not

File: Library/Formula/delta.rb (hunk 2 above)
Comment (DomT4) [inline Library/Formula/delta.rb]: ` https://mirrors.kernel.org/debian/pool/main/d/delta/delta_2006.08.03.orig.tar.gz ` won't mess you around.

Comment (tamird) [inline Library/Formula/delta.rb]: thanks!

**Comment by tamird** [comment]
Looks like creduce fails on `./configure` on mountain lion - is there any way to look at `config.log`? http://bot.brew.sh/job/Homebrew%20Pull%20Requests/24277/version=mountain_lion/testReport/junit/brew-test-bot/mountain_lion/install_creduce/

**Comment by DomT4** [comment]
It doesn't like LLVM, even Homebrew's LLVM, on Mountain Lion.

**Comment by tamird** [comment]
@mikemcquaid added tests!

@DomT4 is it possible to mark formulae as not supporting a particular OS X version?

**Comment by DomT4** [comment]
@tamird It's a little drastic, but ` depends_on :macos => :mavericks ` is the line you're looking for.

File: Library/Formula/delta.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: You can pass the multiple arguments to `bin.install` in that cse.

File: Library/Formula/delta.rb [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    %w[--username guest]
+  end
+end
+
+class Delta < Formula
+  homepage "http://delta.tigris.org/"
+  url "https://mirrors.kernel.org/debian/pool/main/d/delta/delta_2006.08.03.orig.tar.gz"
+  sha256 "38184847a92b01b099bf927dbe66ef88fcfbe7d346a7304eeaad0977cb809ca0"
+
+  head "http://delta.tigris.org/svn/delta/trunk",
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: can you do `guest@delta.` instead?

Comment (tamird) [inline Library/Formula/delta.rb]: nope, I get a 401 :(

Comment (mikemcquaid) [inline Library/Formula/delta.rb]: Let's just remove `head` then, I think.

File: Library/Formula/creduce.rb [hunk 5]
```diff
@@ -0,0 +1,80 @@
+class Creduce < Formula
+  homepage "http://embed.cs.utah.edu/creduce/"
+  url "http://embed.cs.utah.edu/creduce/creduce-2.2.1.tar.gz"
+  sha256 "c2fb1b00feafee20b467674679b68a58bc14556c2f373bb9688150f1fc740072"
+
+  head "https://github.com/rgov/creduce.git"
+
+  # Additional requirements flex and indent should already be available on OS X
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Can remove this comment.

File: Library/Formula/creduce.rb [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+      r.stage do
+        system "perl", "Makefile.PL", "INSTALL_BASE=#{libexec}"
+        system "make"
+        system "make", "install"
+      end
+    end
+
+    system "./configure", "--prefix=#{prefix}",
+                          "--disable-dependency-tracking",
+                          "--with-llvm=#{llvm_prefix}",
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Just inline the `Formula` call here.

File: Library/Formula/creduce.rb [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                          "--bindir=#{libexec}"
+    system "make"
+    system "make", "install"
+
+    (bin + "creduce").write_env_script("#{libexec}/creduce", :PERL5LIB => ENV["PERL5LIB"])
+  end
+
+  test do
+    # look at cmake.rb
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: This will need done before we merge.

File: Library/Formula/delta.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+    #include <stdio.h>
+
+    int main() {
+      int i = -1;
+      unsigned int j = i;
+      printf("%d\n", j);
+    }
+
+    C
+    (testpath/"test1.sh").write(<<-SH)
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: Use `EOS.undent`

File: Library/Formula/delta.rb [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+  head "http://delta.tigris.org/svn/delta/trunk",
+    :using => GuestSubversionDownloadStrategy
+
+  def install
+    system "make"
+    bin.install("delta", "multidelta", "topformflat")
+  end
+
+  test do
+    (testpath/"test1.c").write(<<-C)
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: Use `EOS.undent`

File: Library/Formula/creduce.rb (hunk 5 above)
Comment (tamird) [inline Library/Formula/creduce.rb]: done

File: Library/Formula/creduce.rb (hunk 6 above)
Comment (tamird) [inline Library/Formula/creduce.rb]: done

File: Library/Formula/creduce.rb (hunk 7 above)
Comment (tamird) [inline Library/Formula/creduce.rb]: done

File: Library/Formula/delta.rb (hunk 9 above)
Comment (tamird) [inline Library/Formula/delta.rb]: done

File: Library/Formula/delta.rb (hunk 8 above)
Comment (tamird) [inline Library/Formula/delta.rb]: done

**Comment by tamird** [comment]
@mikemcquaid green!

File: Library/Formula/creduce.rb [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+                          "--bindir=#{libexec}"
+    system "make"
+    system "make", "install"
+
+    (bin + "creduce").write_env_script("#{libexec}/creduce", :PERL5LIB => ENV["PERL5LIB"])
+  end
+
+  test do
+    (testpath/"test1.c").write(<<-EOS.undent)
+    #include <stdio.h>
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: You can indent this one further in (just to treat this like a block.

File: Library/Formula/creduce.rb [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+      printf("%d\n", j);
+    }
+
+    EOS
+    (testpath/"test1.sh").write(<<-EOS.undent)
+    #!/usr/bin/env bash
+
+    clang -Weverything "$(dirname "${BASH_SOURCE[0]}")"/test1.c 2>&1 | \
+    grep 'implicit conversion changes signedness'
+
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: same here.

File: Library/Formula/delta.rb [hunk 12]
```diff
@@ -0,0 +1,33 @@
+class Delta < Formula
+  homepage "http://delta.tigris.org/"
+  url "https://mirrors.kernel.org/debian/pool/main/d/delta/delta_2006.08.03.orig.tar.gz"
+  sha256 "38184847a92b01b099bf927dbe66ef88fcfbe7d346a7304eeaad0977cb809ca0"
+
+  def install
+    system "make"
+    bin.install("delta", "multidelta", "topformflat")
```
Comment (mikemcquaid) [inline Library/Formula/delta.rb]: Don't need the parens here.

File: Library/Formula/creduce.rb [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+                          "--with-llvm=#{Formula["llvm"].prefix}",
+                          "--bindir=#{libexec}"
+    system "make"
+    system "make", "install"
+
+    (bin + "creduce").write_env_script("#{libexec}/creduce", :PERL5LIB => ENV["PERL5LIB"])
+  end
+
+  test do
+    (testpath/"test1.c").write(<<-EOS.undent)
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Don't need parents around the `EOS.undent` (or in any other cases)

File: Library/Formula/creduce.rb [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+    end
+
+    system "./configure", "--prefix=#{prefix}",
+                          "--disable-dependency-tracking",
+                          "--with-llvm=#{Formula["llvm"].prefix}",
+                          "--bindir=#{libexec}"
+    system "make"
+    system "make", "install"
+
+    (bin + "creduce").write_env_script("#{libexec}/creduce", :PERL5LIB => ENV["PERL5LIB"])
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: `bin + "creduce"` can lose the spaces.

File: Library/Formula/creduce.rb [hunk 15]
```diff
@@ -0,0 +1,93 @@
+class Creduce < Formula
+  homepage "http://embed.cs.utah.edu/creduce/"
+  url "http://embed.cs.utah.edu/creduce/creduce-2.2.1.tar.gz"
+  sha256 "c2fb1b00feafee20b467674679b68a58bc14556c2f373bb9688150f1fc740072"
+
+  head "https://github.com/rgov/creduce.git"
+
+  depends_on "astyle"
+  depends_on "delta"
+  depends_on "llvm" => "with-clang"
```
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Can this be coaxed into using the system clang?

Comment (tamird) [inline Library/Formula/creduce.rb]: this formula depends on llvm-config, which is not provided by the system AFAICT. bottling `creduce` would possibly remove the llvm dependency, if it were statically linked

**Comment by tamird** [comment]
@mikemcquaid addressed comments

File: Library/Formula/creduce.rb (hunk 15 above)
Comment (mikemcquaid) [inline Library/Formula/creduce.rb]: Is it statically linked? What's it using `llvm-config` for, I wonder? I wonder if we can provide that information some other way?

Comment (tamird) [inline Library/Formula/creduce.rb]: Doesn't look like it's statically linked. It's used in: https://github.com/csmith-project/creduce/blob/creduce-2.2.1/clang_delta/Makefile.am#L22:L27. Do you know how we might provide that information another way?

The results on my machine:
```shell
$ ./Cellar/llvm/3.5.1/bin/llvm-config --cxxflags
-I/usr/local/Cellar/llvm/3.5.1/include  -D_DEBUG -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -O3  -std=c++11 -fvisibility-inlines-hidden -fno-exceptions -fno-common -Woverloaded-virtual -Wcast-qual
$ ./Cellar/llvm/3.5.1/bin/llvm-config --ldflags
-L/usr/local/Cellar/llvm/3.5.1/lib
$ ./Cellar/llvm/3.5.1/bin/llvm-config --includedir
/usr/local/Cellar/llvm/3.5.1/include
$ ./Cellar/llvm/3.5.1/bin/llvm-config --libs
-lLLVMLTO -lLLVMObjCARCOpts -lLLVMLinker -lLLVMipo -lLLVMVectorize -lLLVMBitWriter -lLLVMIRReader -lLLVMAsmParser -lLLVMTableGen -lLLVMDebugInfo -lLLVMOption -lLLVMX86Disassembler -lLLVMX86AsmParser -lLLVMX86CodeGen -lLLVMSelectionDAG -lLLVMAsmPrinter -lLLVMX86Desc -lLLVMX86Info -lLLVMX86AsmPrinter -lLLVMX86Utils -lLLVMJIT -lLLVMLineEditor -lLLVMMCAnalysis -lLLVMMCDisassembler -lLLVMInstrumentation -lLLVMInterpreter -lLLVMCodeGen -lLLVMScalarOpts -lLLVMInstCombine -lLLVMTransformUtils -lLLVMipa -lLLVMAnalysis -lLLVMProfileData -lLLVMMCJIT -lLLVMTarget -lLLVMRuntimeDyld -lLLVMObject -lLLVMMCParser -lLLVMBitReader -lLLVMExecutionEngine -lLLVMMC -lLLVMCore -lLLVMSupport
$ ./Cellar/llvm/3.5.1/bin/llvm-config --system-libs
-lz -lpthread -ledit -lcurses -lm
```

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

For future reference we prefer only one formula changed per commit.

**Pull request closed without merging**